File: HandleUtils/ObjectDirectory.cpp
//  Copyright 2015 Google Inc. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http ://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#include "stdafx.h"
#include "ObjectDirectory.h"
#include "ObjectDirectoryEntry.h"
#include "WindowsInternals.h"
#include "HandleUtils.h"
#include "ScopedHandle.h"
#include "typed_buffer.h"
#include "TokenLibrary.h"

#include <vcclr.h>
#include <memory>
#include <sddl.h>

namespace HandleUtils {

  using namespace System;
  using namespace System::Collections::Generic;
  using namespace System::Security::Principal;

  struct LocalFreeDeleter
  {
    typedef void* pointer;
    void operator()(void* p) {
      ::LocalFree(p);
    }
  };

  class BoundaryDescriptor
  {
  public:
    BoundaryDescriptor(String^ name)
      : boundary_desc_(nullptr) {
      pin_ptr<const wchar_t> pname = PtrToStringChars(name);
      boundary_desc_ = ::CreateBoundaryDescriptorW(pname, 0);
      if (!boundary_desc_)
        throw gcnew System::ComponentModel::Win32Exception();
    }

    ~BoundaryDescriptor() {
      if (boundary_desc_) {
        DeleteBoundaryDescriptor(boundary_desc_);
      }
    }

    void AddSid(String^ sid)
    {      
      pin_ptr<const wchar_t> psid = PtrToStringChars(sid);
      PSID p;      
      if (!::ConvertStringSidToSid(psid, &p))
        throw gcnew System::ComponentModel::Win32Exception();
      std::unique_ptr<void, LocalFreeDeleter> sid_buf = nullptr;
      sid_buf.reset(p);

      SID_IDENTIFIER_AUTHORITY il_id_auth = { {0,0,0,0,0,0x10} };      
      PSID_IDENTIFIER_AUTHORITY sid_id_auth = GetSidIdentifierAuthority(p);

      if (memcmp(il_id_auth.Value, sid_id_auth->Value, sizeof(il_id_auth.Value)) == 0)
      {
        if (!::AddIntegrityLabelToBoundaryDescriptor(&boundary_desc_, p))
          throw gcnew System::ComponentModel::Win32Exception();
      }
      else
      {
        if (!AddSIDToBoundaryDescriptor(&boundary_desc_, p))
          throw gcnew System::ComponentModel::Win32Exception();
      }
    }
    
    HANDLE boundry_desc() {
      return boundary_desc_;
    }

  private:
    HANDLE boundary_desc_;
  };

  NativeHandle^ ObjectDirectory::OpenPath(ObjectDirectory^ root, System::String^ path)
  {    
    ScopedHandle obj_dir;
    
    OBJECT_ATTRIBUTES_WITH_NAME obj_attr(path, OBJ_CASE_INSENSITIVE, 
      root != nullptr ? root->Handle->DangerousGetHandle().ToPointer() : nullptr);

    DEFINE_NTDLL(NtOpenDirectoryObject);

    NTSTATUS status = fNtOpenDirectoryObject(obj_dir.GetBuffer(), MAXIMUM_ALLOWED, &obj_attr);

    if (!NT_SUCCESS(status))
    {
      throw gcnew System::ComponentModel::Win32Exception(NtStatusToWin32(status));
    }

    return obj_dir.DetachAsNativeHandle();
  }

  NativeHandle^ ObjectDirectory::OpenNamespace(System::String^ path)
  {
    array<String^>^ parts = path->Split('@', 2);
    String^ obj_name = parts->Length > 1 ? parts[1] : parts[0];

    BoundaryDescriptor boundary(obj_name);

    if (parts->Length > 1)
    {
      for each (String^ sid in parts[0]->Split(':'))
      {
        boundary.AddSid(sid);
      }
    }

    String^ prefix = Guid::NewGuid().ToString();
    pin_ptr<const wchar_t> pprefix = PtrToStringChars(prefix);

    ScopedHandle ns(::OpenPrivateNamespaceW(boundary.boundry_desc(), pprefix), false);
    if (!ns.IsValid())
      throw gcnew System::ComponentModel::Win32Exception();
    
    return ns.DetachAsNativeHandle();
  }

	void ObjectDirectory::PopulateEntries()
	{		
		bool readacl = true;
		this->_entries = gcnew List<ObjectDirectoryEntry^>();    
    
    unsigned int granted_access = NativeBridge::GetGrantedAccess(_handle);

		if ((granted_access & READ_CONTROL) == READ_CONTROL)
		{
			_sd = GetSecurityDescriptor(_handle->DangerousGetHandle().ToPointer());
			_sddl = NativeBridge::GetStringSecurityDescriptor(_sd);
		}
		else
		{
			_sd = gcnew array<unsigned char>(0);
			_sddl = "";
		}

		_full_path = QueryObjectName(_handle->DangerousGetHandle().ToPointer());
		if (String::IsNullOrWhiteSpace(_full_path))
		{
			_full_path = _orig_path;
		}

    if ((granted_access & DIRECTORY_QUERY) != DIRECTORY_QUERY)
      return;

		ULONG context = 0;
		typed_buffer_ptr<DIRECTORY_BASIC_INFORMATION> dir_info(2048);
		ULONG length = 0;
    NTSTATUS status = 0;

    DEFINE_NTDLL(NtQueryDirectoryObject);

		while ((status = fNtQueryDirectoryObject(_handle->DangerousGetHandle().ToPointer(), 
      dir_info, (ULONG)dir_info.size(),
			TRUE, FALSE, &context, &length)) != STATUS_NO_MORE_ENTRIES)
		{
			if (!NT_SUCCESS(status))
			{
				throw gcnew System::ComponentModel::Win32Exception(NtStatusToWin32(status));
			}

			ObjectDirectoryEntry^ entry = gcnew ObjectDirectoryEntry(UnicodeNameToString(dir_info->ObjectName), UnicodeNameToString(dir_info->ObjectTypeName), this);

			this->_entries->Add(entry);
		}
	}

	ObjectDirectory::ObjectDirectory(ObjectDirectory^ root, System::String^ name)
	{		
    _orig_path = name;

    if (this->_orig_path->StartsWith("\\") || root != nullptr)
    {
      _handle = OpenPath(root, this->_orig_path);
    }
    else
    {
      _handle = OpenNamespace(this->_orig_path);
    }

		PopulateEntries();
	}

  ObjectDirectory^ ObjectDirectory::Duplicate()
  {
    ObjectDirectory^ ret = gcnew ObjectDirectory();
    ret->_sddl = _sddl;
    ret->_sd = (array<unsigned char>^)_sd->Clone();
    ret->_orig_path = _orig_path;
    ret->_full_path = _full_path;
    ret->_entries = gcnew  System::Collections::Generic::List<ObjectDirectoryEntry^>();
    for each(ObjectDirectoryEntry^ entry in _entries)
    {
      ret->_entries->Add(gcnew ObjectDirectoryEntry(entry->ObjectName, entry->TypeName, ret));
    }
    ret->_handle = _handle->Duplicate();
    return ret;
  }

}

File: ManagedHandleUtils/ObjectDirectory.cs
//  Copyright 2016 Google Inc. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

using NtApiDotNet;
using System;
using System.Collections.Generic;
using System.Linq;

namespace SandboxAnalysisUtils
{
    public sealed class ObjectDirectory : IDisposable
    {
        private string _orig_path;
        private string _full_path;
        private List<ObjectDirectoryEntry> _entries;
        private string _sddl;
        private byte[] _sd;
        private NtDirectory _directory;

        private void PopulateEntries()
        {
            DirectoryAccessRights granted_access = _directory.GrantedAccess;
            if ((granted_access & DirectoryAccessRights.ReadControl) == DirectoryAccessRights.ReadControl)
            {
                _sd = _directory.GetSecurityDescriptorBytes(SecurityInformation.Dacl | SecurityInformation.Label | SecurityInformation.Group | SecurityInformation.Owner);
                _sddl = NtSecurity.SecurityDescriptorToSddl(_sd, SecurityInformation.Dacl | SecurityInformation.Label | SecurityInformation.Group | SecurityInformation.Owner);
            }
            else
            {
                _sd = new byte[0];
                _sddl = String.Empty;
            }


            _full_path = _directory.FullPath;
            if (String.IsNullOrWhiteSpace(_full_path))
            {
                _full_path = _orig_path;
            }

            if ((granted_access & DirectoryAccessRights.Query) != DirectoryAccessRights.Query)
            {
                _entries = new List<ObjectDirectoryEntry>();
            }
            else
            {
                _entries = new List<ObjectDirectoryEntry>(_directory.Query().Select(e => new ObjectDirectoryEntry(e.Name, e.TypeName, this)));
            }
        }

        public ObjectDirectory Duplicate()
        {
            ObjectDirectory ret = new ObjectDirectory();
            ret._sddl = _sddl;
            ret._sd = (byte[])_sd.Clone();
            ret._orig_path = _orig_path;
            ret._full_path = _full_path;
            ret._entries = new List<ObjectDirectoryEntry>(_entries.Select(e => new ObjectDirectoryEntry(e.ObjectName, e.TypeName, ret)));
            ret._directory = _directory.Duplicate();
            return ret;
        }

        public void Refresh()
        {
            PopulateEntries();
        }

        public string FullPath
        {
            get
            {
                return _full_path;
            }
        }

        public string OriginalPath
        {
            get
            {
                return _orig_path;
            }
        }

        public byte[] SecurityDescriptor
        {
            get
            {
                return (byte[])_sd.Clone();
            }
        }

        public string StringSecurityDescriptor
        {
            get
            {
                return _sddl;
            }
        }

        public ObjectDirectory ParentDirectory
        {
            get
            {
                int index = _full_path.LastIndexOf("\\");
                if (index > 0)
                {
                    return new ObjectDirectory(null, _full_path.Substring(0, index));
                }
                else
                {
                    return null;
                }
            }
        }

        public string Name
        {
            get
            {
                int index = _full_path.LastIndexOf("\\");
                if (index > 0)
                {
                    return _full_path.Substring(index + 1);
                }
                else
                {
                    return _full_path;
                }
            }
        }

        public IEnumerable<ObjectDirectoryEntry> Entries
        {
            get
            {
                if (_entries == null)
                {
                    PopulateEntries();
                }
                return _entries.AsReadOnly();
            }
        }
        

        public NtDirectory Directory
        {
            get { return _directory; }
        }

        private static NtDirectory OpenPath(ObjectDirectory root, string path)
        {
            return NtDirectory.Open(path, root != null ? root._directory : null, DirectoryAccessRights.MaximumAllowed);
        }

        private static NtDirectory OpenNamespace(string path)
        {
            using (BoundaryDescriptor boundary = BoundaryDescriptor.CreateFromString(path))
            {
                return NtDirectory.OpenPrivateNamespace(boundary);
            }
        }

        internal ObjectDirectory(ObjectDirectory root, string object_path)
        {
            _orig_path = object_path;

            if (_orig_path.StartsWith("\\") || root != null)
            {
                _directory = OpenPath(root, _orig_path);
            }
            else
            {
                _directory = OpenNamespace(_orig_path);
            }

            PopulateEntries();
        }

        private ObjectDirectory()
        { }

        #region IDisposable Support
        private bool disposedValue = false; // To detect redundant calls

        private void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (_directory != null)
                {
                    _directory.Close();
                    _directory = null;
                }
                disposedValue = true;
            }
        }

        ~ObjectDirectory()
        {
            Dispose(false);
        }

        // This code added to correctly implement the disposable pattern.
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        #endregion
    }
}
